lib/charms/lte_core_interface/v0/lte_core_interface.py

```diff
@@ -0,0 +1,245 @@
+        self.charm = charm
+        self.relationship_name = relationship_name
+        self.framework.observe(
+            charm.on[relationship_name].relation_changed, self._on_relation_changed
+        )
+
+    @staticmethod
+    def _relation_data_is_valid(remote_app_relation_data: dict) -> bool:
+        try:
+            validate(instance=remote_app_relation_data, schema=REQUIRER_JSON_SCHEMA)
+            return True
+        except exceptions.ValidationError as e:
+            logger.error("Invalid relation data: %s", e)
+            return False
+
+    def _on_relation_changed(self, event: RelationChangedEvent) -> None:
+        """Handler triggered on relation changed event.
+
+        Args:
+            event: Juju event (RelationChangedEvent)
+
+        Returns:
+            None
+        """
+        relation = self.model.get_relation(self.relationship_name)
+        if not relation:
+            logger.warning("No relation with name %s", self.relationship_name)
+            return
+        if not relation.app:
+            logger.warning("No remote application in relation: %s", self.relationship_name)
+            return
+        remote_app_relation_data = relation.data[relation.app]
+        if not self._relation_data_is_valid(dict(remote_app_relation_data)):
+            logger.warning(
+                "Provider relation data did no pass JSON Schema validation: \n%s",
+                event.relation.data[event.app],  # type: ignore[index]
+            )
+            return
+        self.on.lte_core_available.emit(
+            mme_ipv4_address=remote_app_relation_data["mme_ipv4_address"],
+        )
+
+
+class CoreProvides(Object):
+    """Class to be instantiated by the charm providing the core."""
+
+    def __init__(self, charm: CharmBase, relationship_name: str):
+        """Init."""
+        super().__init__(charm, relationship_name)
+        self.relationship_name = relationship_name
+        self.charm = charm
+
+    @staticmethod
+    def mme_ipv4_address_is_valid(mme_ipv4_address: str) -> bool:
+        """Returns whether mme ipv4 address is valid."""
+        # TODO: Is this approach ok?
+        try:
+            IPv4Address(mme_ipv4_address)
+            return True
+        except:  # noqa: E722
+            return False
+
+    def set_core_information(self, mme_ipv4_address: str) -> None:
+        """Sets mme_ipv4_address in the application relation data.
+
+        Args:
+            mme_ipv4_address: MME ipv4 address
+        Returns:
+            None
+        """
+        if not self.charm.unit.is_leader():
+            raise RuntimeError("Unit must be leader to set application relation data.")
+        if not self.mme_ipv4_address_is_valid(mme_ipv4_address):
+            raise AddressValueError("Invalid MME IPv4 address.")
+        if relation := self.model.get_relation(self.relationship_name):
+            relation.data[self.charm.app].update({"mme_ipv4_address": mme_ipv4_address})
```

Comment:
at this point everything is string, so there are no issues, but if this library would be extended with non-string information you might want to use `json.dumps` here to make sure you are storing a json string.
And `json.loads` on the other side, which also saves you the conversion to dict as it does it :)

Reply:
We have only one field at this point, and I don't expect to require any additional one.